Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,221 @@
+			.flatMap { platform in
+				return self.cacheForPlatform(platform) != nil ? [platform] : []
+			})
+	}
+	
+	func toJSONObject() -> AnyObject {
+		return Platform.supportedPlatforms.reduce([String: AnyObject](), combine: { (dict, platform) in
+			var dict = dict
+			dict[platform.rawValue] = cacheForPlatform(platform)?.toJSONObject()
+			return dict
+		})
+	}
+}
+
+extension VersionFile: Decodable {
+	static func decode(j: JSON) -> Decoded<VersionFile> {
+		return curry(self.init)
+			<^> j <|? Platform.Mac.rawValue
+			<*> j <|? Platform.iOS.rawValue
+			<*> j <|? Platform.watchOS.rawValue
+			<*> j <|? Platform.tvOS.rawValue
+	}
+}
+
+private typealias VersionFileData = [String: [String: AnyObject]]
+
+/// Creates a version file for the current dependency in the
+/// Carthage/Build directory which associates its commitish with
+/// the SHA1s of the built frameworks for each platform in order
+/// to allow those frameworks to be skipped in future builds.
+///
+/// Returns true if the version file was successfully created.
+public func createVersionFileForDependency(dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, buildProductURLs: [NSURL], rootDirectoryURL: NSURL) -> Bool {
+	var frameworksByPlatform: [String: [CachedFramework]] = [:]
+	
+	for url in buildProductURLs {
+		guard let platformName = url.URLByDeletingLastPathComponent?.lastPathComponent else { return false }
+		guard let frameworkName = url.URLByDeletingPathExtension?.lastPathComponent else { return false }
+		
+		let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
+		guard NSFileManager.defaultManager().fileExistsAtPath(frameworkURL.path!) else { return false }
+		guard let sha1 = sha1ForFileAtURL(frameworkURL) else { return false }
+		
+		let cachedFramework = CachedFramework(name: frameworkName, sha1: sha1)
+		
+		var frameworks = frameworksByPlatform[platformName] ?? []
+		frameworks.append(cachedFramework)
+		frameworksByPlatform[platformName] = frameworks
+	}
+	
+	var cachedPlatforms: [String: CachedPlatform] = frameworksByPlatform.reduce([:]) { (cachedPlatforms, entry: (platformName: String, cachedFrameworks: [CachedFramework])) in
+		var cachedPlatforms = cachedPlatforms
+		cachedPlatforms[entry.platformName] = CachedPlatform(commitish: dependency.version.commitish, cachedFrameworks: entry.cachedFrameworks)
+		return cachedPlatforms
+	}
+	
+	let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).URLByResolvingSymlinksInPath!
+	let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+	
+	if let oldVersionFile: VersionFile = readVersionFileAtURL(versionFileURL) {
+		for platform in Platform.supportedPlatforms {
+			if let oldCachedPlatform = oldVersionFile.cacheForPlatform(platform) where cachedPlatforms[platform.rawValue] == nil {
+				cachedPlatforms[platform.rawValue] = oldCachedPlatform
+			}
+		}
+	}
```

Comment:
If you don't cache per platform you blow away state when using the '--platform' flag. For example if you first build with '--platform ios' then later '--platform mac', then build with '--platform ios' again, it won't be cached. 

Additionally you would still the platform variance somewhere, as the SHA's will be different / there may be different counts of frameworks in the different platforms (e.g., they have an ios only target and a mac os only target).